Mathlib/Order/Normal.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Order.SuccPred.InitialSeg
+
+/-!
+# Normal functions
+
+A normal function between well-orders is a strictly monotonic continuous function. Normal functions
+arise chiefly in the context of cardinal and ordinal-valued functions.
+
+We opt for an equivalent definition that's both simpler and often more convenient: a normal function
+is a strictly monotonic function `f` such that at successor limits `a`, `f a` is the least upper
+bound of `f b` with `b < a`.
+
+## TODO
+
+* Prove the equivalence with the standard definition (in some other file).
+* Replace `Ordinal.IsNormal` by this more general notion.
+-/
+
+open Order Set
+
+variable {α β γ : Type*} {a b : α} {f : α → β} {g : β → γ}
+
+namespace Order
+
+/-- A normal function between well-orders is a strictly monotonic continuous function. -/
+structure IsNormal [LinearOrder α] [LinearOrder β] (f : α → β) : Prop where
+  strictMono : StrictMono f
+  isLUB_image_Iio_of_isSuccLimit {a : α} (ha : IsSuccLimit a) : IsLUB (f '' Iio a) (f a)
+
+namespace IsNormal
+
+section LinearOrder
+variable [LinearOrder α] [LinearOrder β] [LinearOrder γ]
+
+/-- This condition is the LHS of the `IsLUB (f '' Iio a) (f a)` predicate. -/
+theorem of_mem_lowerBounds_upperBounds {f : α → β} (hf : StrictMono f)
+    (H : ∀ a, IsSuccLimit a → f a ∈ lowerBounds (upperBounds (f '' Iio a))) : IsNormal f := by
+  refine ⟨hf, @fun a ha ↦ ⟨?_, H _ ha⟩⟩
+  rintro - ⟨b, hb, rfl⟩
+  exact (hf hb).le
+
+theorem of_succ_lt [SuccOrder α] [WellFoundedLT α]
+    (hs : ∀ a, f a < f (succ a)) (hl : ∀ {a}, IsSuccLimit a → IsLUB (f '' Iio a) (f a)) :
+    IsNormal f := by
+  refine ⟨fun a b ↦ SuccOrder.limitRecOn b ?_ ?_ ?_, hl⟩
```

Comment:
Optional, but how does this proof look if you do it with `induction b using SuccOrder.limitRecOn`? From the current structure, it looks like this could be nicer to read...

Reply:
It's a bit nicer. We really should rename the hypotheses in that theorem, but that's a future PR.